_posts/2021-04-28-#17331.md

```diff
@@ -0,0 +1,111 @@
+---
+layout: pr
+date: 2021-04-28
+title: "Use effective values throughout coin selection"
+pr: 17331
+authors: [achow101]
+components: ["wallet"]
+host: Xekyo
+status: upcoming
+commit: 4ac1add
+---
+
+## Notes
+
+- Wallets maintain their balance as a pool of Unspent Transaction Outputs
+  (UTXOs).
+
+- _Coin selection_ refers to the process of picking UTXOs from the wallet's UTXO
+  pool to fund a transaction.
+
+- Beyond the primary goal of funding a transaction, the secondary goals of coin
+  selection are:
+
+    - Minimizing short term and long term fees:  
+      We want to both pay the least right now, but we also want to be good
+      stewards of our UTXO pool and not rack up large UTXO counts that will
+      translate to huge future transaction fees.
+
+    - Maintaining financial privacy:  
+      Ideally, our transactions leak as little information about ourselves and
+      our business partners as possible.
+
+    - Reliably move the payment(s) to finalization:  
+      Reduce the chance of transactions being conflicted and transfer value in a
+      limited time-frame
+
+- Using unconfirmed inputs can make transactions unreliable. When a foreign
+  input is used, the parent transaction may time out or get replaced. Even using
+  self-sent unconfirmed funds may delay the new transaction if the parent
+  transaction has an extensive ancestry, is extraordinarily large, or used a
+  lower feerate than targeted for the child transaction.
+
+- When a new transaction is built, the target feerate is provided as an input.
+
+- When a UTXO is selected as an input, the size of the transaction is increased
+  which in turn increases the necessary fees to maintain the targeted feerate.
+  We can preempt the cost added by the input by considering each UTXO at its
+  _effective value_, its value reduced by the UTXO's input cost at the given
+  feerate.
+
+- Using _effective values_ in coin selection means that we can freely add and
+  remove inputs until we have have raised enough funds to pay for the
+  transaction's fixed costs: the payment amounts, the transaction header and the
+  output data. Whatever final input set we settle for, the increased transaction
+  size and input costs have already been accounted for.
+
+- The _knapsack solver_ sorts the available UTXOs by effective value, and then
+  combines candidate input sets by randomly selecting or skipping inputs until
+  it has sufficient funds for the transaction. When it finds a viable solution,
+  it keeps the solution if it produces a smaller change than the prior stored
+  solution. Then, the algorithm deselects the last input and continues
+  traversing the UTXO list as described, trying to converge on the smallest
+  change output above `MIN_CHANGE`. This random process is repeated up to 1,000
+  times.
+
+- The _Branch and Bound solver (BnB)_ deterministically searches the complete
```

Comment:
I think you should mention the no-change part in this first sentence, because it's a crucial poitn in this PR.

Reply:
Good point, moved it up.